### DISCUSSION TITLE
OpenSSL API and OpenSSH ML-KEM support

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#26073] [General Discussion]
Dear colleagues, OpenSSH has recently implemented support of a hybrid ML-KEM KEX (https://github.com/openssh/openssh-portable/blob/master/kexmlkem768x25519.c) according to https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke-03

In some reasonable future we plan using OpenSSL API for key exchange, both in the downstream version of OpenSSH and upstream version of libssh. I wonder if the OpenSSL EVP API is enough for this as of now or some new API should be used

**Comment by beldmit** [comment]
Ping @baentsch  @vdukhovni

**Comment by mattcaswell** [comment]
The described scheme seems to me very similar to the TLS hybrid scheme, and is based on a simple concatenation of the various values. As such it should be able to use #25990. The hybrid combination is exposed as just another KEM and I see no reason why the existing EVP KEM API would not be sufficient.

**Reply by baentsch** [reply]
Agree with @mattcaswell . Even before generic hybrids are implemented in default provider this particular hybrid is available via oqsprovider in the EVP API, so you should be able to test it already @beldmit and let us know any shortcomings that could inform required improvements for #25990 (or openssl core).

**Reply by baentsch** [reply]
A recent discussion [elsewhere](https://github.com/open-quantum-safe/oqs-provider/issues/579) seems to indicate that the logic proposed in the "TLS1.3 hybrid" and "OpenSSH hybrid" is different in their respective IETF drafts. While this warrants different code (change concatenation order and optionally add a hash) for enabling suitable (differently named) hybrids I still don't see any of this warranting new APIs. What does make sense for sure is add the draft spec reference above to https://github.com/openssl/openssl/issues/25884.

**Comment by petrovr** [comment]
From my point of existing "Encapsulation API" is enough. 

[HYBRID-DEMO](https://gitlab.com/secsh/pkixssh/-/commits/hybrid_demo) is a PKIX-SSH demo implementation on all three KEX described in internet draft:
1. [mlkem768x25519-sha256](https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke/#name-mlkem768x25519-sha256)
2. [mlkem1024nistp384-sha384](https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke/#name-mlkem1024nistp384-sha384) 
3. [mlkem768x25519-sha256](https://datatracker.ietf.org/doc/html/draft-kampanakis-curdle-ssh-pq-ke/#name-mlkem768x25519-sha256)

If uses functionality available from oqs-provider. 

Note key exchange in PKIX-SSH is PKEY based and demo uses ...keygen..., ...encapsulate..., ...decapsulate... OpenSSL API. 
Public key for encapsulation uses function EVP_PKEY_new_raw_public_key_ex(3).


For protocol OpenSSH implements only (3).

**Reply by vdukhovni** [reply]
> Note key exchange in PKIX-SSH is PKEY based and demo uses ...keygen..., ...encapsulate..., ...decapsulate... OpenSSL API.
> Public key for encapsulation uses function EVP_PKEY_new_raw_public_key_ex(3).

Thanks for the doc reference, I see that document needs a small update as part of [merging ML-KEM support](https://github.com/openssl/openssl/pull/26054)
